### ISSUE TITLE
"localhost" is not 127.0.0.1 in freebsd jails

### ISSUE THREAD

**Issue by bagder** [curl/curl#7160]
In regards to #7039 and [curl localhost as a local host](https://daniel.haxx.se/blog/2021/05/31/curl-localhost-as-a-local-host/)

Selected quotes from an IRC conversation I had on June 1st 2021 with Thomas Steen Rasmussen:

- localhost in jails has always been mapped to the jail's IP or you hit the hosts 127.0.0.1 which is usually not what you want
- a jail is like a lightweight VM, when you are SSH or otherwise logged into it it feels like a full machine, which is why it would be surprising if "localhost" didn't work
- the kernel intercepts the lookup and returns the jails own IP instead of 127.0.0.1
- all the software which just connects to localhost by default just works like it would on a real host or another OS
- vnet jails do away with all this (they have a fully virtualised network stack and their own 127.0.0.1), and in 5-10 years non-vnet jails might be a thing of the past. But for now they are very much alive and kicking :)

### curl/libcurl version

git master

### operating system

FreeBSD. Others?

**Comment by jay** [comment]
Frankly this seems like a good argument to not hard code it, I wouldn't have thought of that.

**Comment by bagder** [comment]
I'm not convinced, as we're putting one series of benefits vs a rather niche side-effect. I would've appreciated someone to argue for it here to help us assess what this means to FreeBSD users and possibly others.

**Comment by dch** [comment]
I'm a FreeBSD jails user, and this change doesn't break jails, to my
knowledge. I tried my jail setup (which has this "fake loopback" IP
config) and report no change in expected behaviour before & after the
patch.

If other users can explain their use case more fully, perhaps we can
understand better what the issue is, in practice.

## TLDR

When a jail has a loopback IP bound that is not 127.0.0.1, inside
the jail, `curl localhost` works as you'd expect, before & after
this patch. It doesn't break the kernel's magical re-routing of
127.0.0.1 loopback IP to the actual jail IP.

results below, first using "normal curl", then again with curl+7039

### outside the jail

```
### the non-127.0.0.1 IPv4 assigned as primary IP
# jls
   JID  IP Address      Hostname                      Path
    83  100.64.34.78    curl.skunkwerks.at            /jails/curl
```

### inside the jail, NB lo0 has no IP assigned, lo1 is the jail IP

```
# ifconfig
igb0: flags=8963<UP,BROADCAST,RUNNING,PROMISC,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=4e503bb<RXCSUM,TXCSUM,...>
        ether ac:1f:6b:67:e1:38
        media: Ethernet autoselect (1000baseT <full-duplex>)
        status: active
        nd6 options=23<PERFORMNUD,ACCEPT_RTADV,AUTO_LINKLOCAL>
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=680003<RXCSUM,TXCSUM,LINKSTATE,RXCSUM_IPV6,TXCSUM_IPV6>
        groups: lo
lo1: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=680003<RXCSUM,TXCSUM,LINKSTATE,RXCSUM_IPV6,TXCSUM_IPV6>
        inet 100.64.34.78 netmask 0xffffffff
        groups: lo
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
...

### what happens when we ping a loopback IP?
# ping localhost
PING localhost (127.0.0.1): 56 data bytes
64 bytes from 127.0.0.1: icmp_seq=0 ttl=64 time=0.066 ms
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.080 ms
^C
--- localhost ping statistics ---
2 packets transmitted, 2 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 0.066/0.073/0.080/0.007 ms

### aah yes but will it curl?
# curl -4vsSo /dev/null localhost
*   Trying 127.0.0.1:80...
* Connected to localhost (100.64.34.78) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.76.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< content-length: 0
< location: https://localhost/
<
* Connection #0 to host localhost left intact

### try it with a specific IP
# curl -4vsSo /dev/null 127.0.0.1 -Hhost:localhost
*   Trying 127.0.0.1:80...
* Connected to 127.0.0.1 (100.64.34.78) port 80 (#0)
> GET / HTTP/1.1
> Host:localhost
> User-Agent: curl/7.76.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< content-length: 0
< location: https://localhost/
<
* Connection #0 to host 127.0.0.1 left intact
```

## use patched curl 7.77.0 + PR#7039

```
# pkg install -r skunkwerks curl
Updating skunkwerks repository catalogue...
[curl.skunkwerks.at] Fetching meta.conf: 100%    163 B   0.2kB/s    00:01
[curl.skunkwerks.at] Fetching packagesite.txz: 100%  131 KiB 134.7kB/s    00:01
Processing entries: 100%
skunkwerks repository update completed. 525 packages processed.
All repositories are up to date.
The following 1 package(s) will be affected (of 0 checked):

Installed packages to be UPGRADED:
        curl: 7.76.1 -> 7.77.0 [skunkwerks]

Number of packages to be upgraded: 1

1 MiB to be downloaded.

[curl.skunkwerks.at] [1/1] Fetching curl-7.77.0.txz: 100%    1 MiB   1.4MB/s    00:01
Checking integrity... done (0 conflicting)
[curl.skunkwerks.at] [1/1] Upgrading curl from 7.76.1 to 7.77.0...
[curl.skunkwerks.at] [1/1] Extracting curl-7.77.0: 100%

### try the obvious thing, does it break?

# curl -4vsSo /dev/null localhost
*   Trying 127.0.0.1:80...
* Connected to localhost (100.64.34.78) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.77.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< content-length: 0
< location: https://localhost/
<
* Connection #0 to host localhost left intact


### try connecting directly to IP and set host header
# curl -4vsSo /dev/null 127.0.0.1 -Hhost:localhost
*   Trying 127.0.0.1:80...
* Connected to 127.0.0.1 (100.64.34.78) port 80 (#0)
> GET / HTTP/1.1
> Host:localhost
> User-Agent: curl/7.77.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< content-length: 0
< location: https://localhost/
<
* Connection #0 to host 127.0.0.1 left intact
```

# details, details, details

This seems like a sensible change, and I don't see what the fuss
here is about. The proposed change fixes real-world problems
that are abused to exploit systems and users, and brings curl's
resolver into line with an RFC from 2013, that requires localhost to
be .... local aka loopback addresses.


localhost as localhost will still work in jails as it did before. If
admins have set up jails in such a way that localhost gets a
specific non 127.0.0.0/8 or ::1/8 IP *inside* the jail, then it's
reasonable to assume they are aware of the implications for
this, and are also capable of work-arounds.

Maybe people have a specific use case or situation that I'm not aware
of, so here's mine for comparison. [jz](https://git.io/JfHOu) is a short
shell script that sets up a pretty standard jail, outputs settings for
network etc as it goes along, and finally drops you in via tmux inside
the jail:

- uses own container for its chrooted filesystem
- uses cloned interface for the jail's loopback IPv4, but assigned to a
  separate IP subnet (the "non-127.0.01" localhost, as seen from outside)
- uses a separate interface IPv6 to link jails across servers via VPN
- both interfaces are shared between jails, but the IPs assigned are
  not typical loopback addresses. Within the jail, they work as expected.
- normal kernel routing & firewalling are handled by the host

> jz is a stripped down version of a much larger jail infrastructure,
> where we have 1000s per server, but the functionality vis-a-vis jails
> and networking is effectively the same.

```shell
## initialise a jail called "curl"
$ jz curl
name=curl
## the jail command actually shells out to ifconfig(8) to set up networking
run command: /sbin/ifconfig lo1 inet 100.64.34.78 netmask 255.255.255.255 alias
run command: /sbin/ifconfig zt1flo98dm17np8 inet6 fc7b:c4d6:6be2:8e50:6c98::224e prefixlen 128 alias
## the kernel initialises the jail for us
jail_set(JAIL_CREATE) name=curl
  path=/jails/curl
  host.domainname=skunkwerks.at
  host.hostname=curl.skunkwerks.at
  ip4.addr=100.64.34.78
  ip6.addr=fc7b:c4d6:6be2:8e50:6c98::224e
   ...
   osrelease=13.0-RELEASE
created

run command in jail: /bin/sh /etc/rc
ELF ldconfig path: /lib /usr/lib /usr/lib/compat /usr/local/lib /usr/local/lib/compat/pkg /usr/local/lib/compat/pkg
32-bit compatibility ldconfig path: /usr/lib32
Clearing /tmp (X related).
Updating /var/run/os-release done.
Creating and/or trimming log files.
Updating motd:.
Starting cron.

Wed Jun  2 08:32:17 UTC 2021
run command in jail: /bin/sh -c env -i TERM=xterm-256color tmux -u new-session -DAs default
```

### inside the jail

I think this is pretty common for almost everybody:

```
root@curl:/ # sysctl security.jail.jailed
security.jail.jailed: 1

root@curl:/ # hostname
curl.skunkwerks.at

root@curl:/ # egrep -v '^#' /etc/hosts
::1                     localhost localhost.skunkwerks.at
127.0.0.1               localhost localhost.skunkwerks.at

## we use a DNS resolver outside the jail (still inside my network)
root@curl:/ # cat /etc/resolv.conf
search skunkwerks.at
nameserver 172.16.1.1
options edns0
```

### network config from inside the jail

```
root@curl:/ # ifconfig
... abridged for relevance
lo1: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=680003<RXCSUM,TXCSUM,LINKSTATE,RXCSUM_IPV6,TXCSUM_IPV6>
        inet 100.64.34.78 netmask 0xffffffff
        groups: lo
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>

zt1flo98dm17np8: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 5000 mtu 2800
        options=80000<LINKSTATE>
        ether 2a:3d:9d:3c:2f:91
        hwaddr 58:9c:fc:10:ff:d1
        inet6 fc7b:c4d6:6be2:8e50:6c98::224e prefixlen 128
        groups: tap
        media: Ethernet autoselect
        status: active
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
        Opened by PID 27368
```

### what do we see using ping today

```
## ping on IP4 "localhost" works fine
root@curl:/ # ping 127.0.0.1
PING 127.0.0.1 (127.0.0.1): 56 data bytes
64 bytes from 127.0.0.1: icmp_seq=0 ttl=64 time=0.064 ms
^C
--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 0.064/0.064/0.064/0.000 ms

## ping on IP6 "localhost" fails as we have *no* ::1/8 alias, only
## private IPv6 range
root@curl:/ # ping ::1
ping: connect() ssend: Can't assign requested address
```

### what we see with curl today

```
### no services listening *within the jail* on port 80
root@curl:/ # sockstat -46lp 80
USER     COMMAND    PID   FD PROTO  LOCAL ADDRESS         FOREIGN ADDRESS

root@curl:/ # curl --version
curl 7.76.1 (amd64-portbld-freebsd13.0) libcurl/7.76.1 OpenSSL/1.1.1k ...
Release-Date: 2021-04-14
Protocols: dict file ftp ftps gopher gophers http https ...
Features: alt-svc AsynchDNS GSS-API HTTP2 HTTPS-proxy IPv6 ...

### but when we "curl loopback" today we get a response
### from the server *outside* the jail listening on port 80
root@curl:/ # curl -4vsSo /dev/null localhost
*   Trying 127.0.0.1:80...
* Connected to localhost (100.64.34.78) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.76.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< content-length: 0
< location: https://localhost/
<
* Connection #0 to host localhost left intact
```

Above we can see the "remapping" that takes place outside the jail,
where our jail's perceived local 127.0.0.1 address is munged by the
kernel, into the first IP address assigned to the jail.

For completeness, here's our external non-jailed web server listening
on all tcp flavours, port 80. This will also encompass port 80 on
the IPs assigned to the jail, but the jail itself doesn't see that.

```
## outside the jail
$ sudo sockstat -l46p 80
USER	COMMAND	PROTO	LOCAL
www	haproxy	tcp46	*:80
```

Using [ngrep] (a more readable tcpdump tool) on the host, we can see the
actual network packets being sent over the loopback interfaces (cloned
or not makes no difference here to the kernel):


```
$ sudo ngrep -qid lo0 -W byline port 80

interface: lo0 (127.0.0.0/255.0.0.0)
filter: (ip or ip6) and ( port 80 )

T 100.64.34.78:6868 -> 100.64.34.78:80 [AP]
GET / HTTP/1.1.
Host: localhost.
User-Agent: curl/7.76.1.
Accept: */*.
.

T 100.64.34.78:80 -> 100.64.34.78:6868 [AP]
HTTP/1.1 301 Moved Permanently.
content-length: 0.
location: https://localhost/.
.
```

Using tcpdump(1) inside the jail we see the same IP used too, if you
don't trust ngrep.

```
### jailed
# tcpdump -i lo0 -vvv port 80
tcpdump: listening on lo0, link-type NULL (BSD loopback), capture size 262144 bytes
09:14:51.017158 IP (tos 0x0, ttl 64, id 63927, offset 0, flags [none], proto TCP (6), length 60, bad cksum 0 (->73e8)!)
    100.64.34.78.16268 > 100.64.34.78.http: Flags [S], cksum 0x0d4b (incorrect -> 0xa63c), seq 2067585655, win 65535, options [mss 16344,nop,wscale 9,sackOK,TS val 3031568446 ecr 0], length 0
09:14:51.017189 IP (tos 0x0, ttl 64, id 33060, offset 0, flags [none], proto TCP (6), length 60, bad cksum 0 (->ec7b)!)
    100.64.34.78.http > 100.64.34.78.16268: Flags [S.], cksum 0x0d4b (incorrect -> 0x1d1b), seq 1186271239, ack 2067585656, win 65535, options [mss 16344,nop,wscale 9,sackOK,TS val 2479922819 ecr 3031568446], length 0
09:14:51.017209 IP (tos 0x0, ttl 64, id 25137, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->b77)!)
    100.64.34.78.16268 > 100.64.34.78.http: Flags [.], cksum 0x0d43 (incorrect -> 0x6601), seq 1, ack 1, win 8204, options [nop,nop,TS val 3031568446 ecr 2479922819], length 0
09:14:51.017251 IP (tos 0x0, ttl 64, id 3049, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->61bf)!)
    100.64.34.78.http > 100.64.34.78.16268: Flags [.], cksum 0x0d43 (incorrect -> 0x6601), seq 1, ack 1, win 8204, options [nop,nop,TS val 2479922819 ecr 3031568446], length 0
09:14:51.017288 IP (tos 0x0, ttl 64, id 62928, offset 0, flags [none], proto TCP (6), length 125, bad cksum 0 (->778e)!)
    100.64.34.78.16268 > 100.64.34.78.http: Flags [P.], cksum 0x0d8c (incorrect -> 0x1018), seq 1:74, ack 1, win 8204, options [nop,nop,TS val 3031568446 ecr 2479922819], length 73: HTTP, length: 73
        GET / HTTP/1.1
        Host: localhost
        User-Agent: curl/7.76.1
        Accept: */*

09:14:51.017309 IP (tos 0x0, ttl 64, id 36192, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->e047)!)
    100.64.34.78.http > 100.64.34.78.16268: Flags [.], cksum 0x0d43 (incorrect -> 0x65b8), seq 1, ack 74, win 8204, options [nop,nop,TS val 2479922819 ecr 3031568446], length 0
09:14:51.017530 IP (tos 0x0, ttl 64, id 61066, offset 0, flags [none], proto TCP (6), length 135, bad cksum 0 (->7eca)!)
    100.64.34.78.http > 100.64.34.78.16268: Flags [P.], cksum 0x0d96 (incorrect -> 0x65be), seq 1:84, ack 74, win 8204, options [nop,nop,TS val 2479922820 ecr 3031568446], length 83: HTTP, length: 83
        HTTP/1.1 301 Moved Permanently
        content-length: 0
        location: https://localhost/

09:14:51.017542 IP (tos 0x0, ttl 64, id 64013, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->739a)!)
    100.64.34.78.16268 > 100.64.34.78.http: Flags [.], cksum 0x0d43 (incorrect -> 0x6563), seq 74, ack 84, win 8204, options [nop,nop,TS val 3031568447 ecr 2479922820], length 0
09:14:51.017761 IP (tos 0x0, ttl 64, id 2711, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->6311)!)
    100.64.34.78.16268 > 100.64.34.78.http: Flags [F.], cksum 0x0d43 (incorrect -> 0x6562), seq 74, ack 84, win 8204, options [nop,nop,TS val 3031568447 ecr 2479922820], length 0
09:14:51.017779 IP (tos 0x0, ttl 64, id 43788, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->c29b)!)
    100.64.34.78.http > 100.64.34.78.16268: Flags [.], cksum 0x0d43 (incorrect -> 0x6562), seq 84, ack 75, win 8204, options [nop,nop,TS val 2479922820 ecr 3031568447], length 0
09:14:51.017800 IP (tos 0x0, ttl 64, id 59742, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->8449)!)
    100.64.34.78.http > 100.64.34.78.16268: Flags [F.], cksum 0x0d43 (incorrect -> 0x6561), seq 84, ack 75, win 8204, options [nop,nop,TS val 2479922820 ecr 3031568447], length 0
09:14:51.017813 IP (tos 0x0, ttl 64, id 16375, offset 0, flags [none], proto TCP (6), length 52, bad cksum 0 (->2db1)!)
    100.64.34.78.16268 > 100.64.34.78.http: Flags [.], cksum 0x0d43 (incorrect -> 0x6562), seq 75, ack 85, win 8203, options [nop,nop,TS val 3031568447 ecr 2479922820], length 0
```

### changing /etc/hosts

Now, let's *modify* `/etc/hosts` to use `100.64.34.78 localhost ...`
instead. I have no idea *why* one would want to do this explicitly, but
... whatever.

```
root@curl:/tmp # grep -v '^#' /etc/hosts
::1                     localhost localhost.skunkwerks.at
100.64.34.78 localhost localhost.skunkwerks.at

root@curl:/tmp # getaddrinfo -f inet localhost
dgram inet udp 100.64.34.78 0
stream inet tcp 100.64.34.78 0
seqpacket inet sctp 100.64.34.78 0

### a non-curl HTTP client that obstinately handles 301 but evidently
### connects to the non-jailed external webserver
root@curl:/tmp # fetch http://localhost/ -o -
SSL certificate subject doesn't match host localhost
fetch: http://localhost/: Authentication error

### curly things
root@curl:/tmp # curl -4vsSo /dev/null localhost
*   Trying 100.64.34.78:80...
* Connected to localhost (100.64.34.78) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.76.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< content-length: 0
< location: https://localhost/
<
* Connection #0 to host localhost left intact
```

The [ngrep] and tcpdump responses are what you'd expect, so I just leave
the more readable ngrep one here:

```
$ sudo ngrep -qid -W byline port 80

T 100.64.34.78:2985 -> 100.64.34.78:80 [AP]
GET / HTTP/1.1.
Host: localhost.
User-Agent: curl/7.76.1.
Accept: */*.
.


T 100.64.34.78:80 -> 100.64.34.78:2985 [AP]
HTTP/1.1 301 Moved Permanently.
content-length: 0.
location: https://localhost/.
.
```

In other words, setting localhost explicitly to a locally bound IP
address that is *not* in the 127.0.0.0/8 net seems to behave perfectly
cromulently.

[ngrep]: https://github.com/jpr5/ngrep

**Comment by xquery** [comment]
the world would be a better place if all issue comments were as lucid as @dch last one was ....

**Comment by jay** [comment]
@bagder does the OP have any response to this?

**Comment by bagder** [comment]
(The only truly valid responses to this subject would be filed in this issue.) @dch has responded to several of the people who yelled the loudest (over on Mastodon) but I've not seen a single legitimate follow-up in any forum that further explains this issue. I'm moving to close this issue within a few days. To me, it seems that the objections were based on misunderstandings and/or complaints about other things in the change that aren't relevant to this issue.

**Comment by bagder** [comment]
Thanks a lot @dch for that very extensive and detailed reply on this subject. Case closed!
